### PR TITLE
refactor: use `LazyLock` in the `user_doc` macro

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/11687

## Rationale for this change
Replace `OnceLock` with `LazyLock`.

## What changes are included in this PR?
Yes.  
Run `./dev/update_function_docs.sh` and verified that the documentation content didn't changed.

## Are these changes tested?
No

### CODE REVIEW

File: datafusion/macros/src/lib.rs [hunk 1]
```diff
@@ -235,27 +248,23 @@ pub fn user_doc(args: TokenStream, input: TokenStream) -> TokenStream {
...
 
-        static #lock_name: OnceLock<Documentation> = OnceLock::new();
-
         impl #name {
-
-                fn doc(&self) -> Option<&Documentation> {
-                    Some(#lock_name.get_or_init(|| {
-                        Documentation::builder(DocSection { include: #doc_section_include, label: #doc_section_lbl, description: #doc_section_description },
+            fn doc(&self) -> Option<&datafusion_doc::Documentation> {
+                static DOCUMENTATION: std::sync::LazyLock<datafusion_doc::Documentation> =
```
Comment (comphead) [inline datafusion/macros/src/lib.rs:256]: thats interesting, I tried doing the same with `OnceLock`, to move it inside the `doc` method but it didn't work. Surprisingly it works with LazyLock

File: datafusion/macros/src/lib.rs [hunk 2]
```diff
@@ -26,16 +26,19 @@ use syn::{parse_macro_input, DeriveInput, LitStr};
 /// declared on `AggregateUDF`, `WindowUDFImpl`, `ScalarUDFImpl` traits.
 ///
 /// Example:
+/// ```ingore
```
Comment (comphead) [inline datafusion/macros/src/lib.rs:29]:
```suggestion
/// ```ignore
```

**Review by comphead — APPROVED** [review]
lgtm thanks @jonahgao 

This PR will be conflicting with https://github.com/apache/datafusion/pull/13685

**Comment by jonahgao** [comment]
Conflicting resolved. Thanks @comphead for the review.

**Pull request merged**